test/tools/ossfuzz/protoToAbiV2.h

```diff
@@ -0,0 +1,249 @@
+
+private:
+	typedef std::vector<std::pair<bool,unsigned>> vecOfBoolUnsignedTy;
+
+	enum class delimiter
+	{
+		ADD,
+		SKIP
+	};
+	enum class calleeType
+	{
+		PUBLIC,
+		EXTERNAL
+	};
+	enum class dataType
+	{
+		BYTES,
+		STRING,
+		VALUE,
+		ARRAY
+	};
+
+	void visit(IntegerType const&);
+	void visit(FixedByteType const&);
+	void visit(AddressType const&);
+	void visit(ArrayType const&);
+	void visit(DynamicByteArrayType const&);
+	void visit(StructType const&);
+	void visit(ValueType const&);
+	void visit(NonValueType const&);
+	void visit(Type const&);
+	void visit(VarDecl const&);
+	void visit(TestFunction const&);
+	void visit(Contract const&);
+	std::string getValueByBaseType(ArrayType const&);
+	void resizeInitArray(
+		ArrayType const& _x,
+		std::string const& _baseType,
+		std::string const& _var
+	);
+	unsigned resizeDimension(
+		bool _isStatic,
+		unsigned _arrayLen,
+		std::string const& _var,
+		std::string const& _type
+	);
+	void resizeHelper(
+		ArrayType const& _x,
+		std::vector<std::string> _arrStrVec,
+		vecOfBoolUnsignedTy _arrInfoVec,
+		std::string const& _var
+	);
+
+	// Utility
+	void appendChecks(dataType _type, std::string const& _varName, std::string const& _rhs);
+	void addVarDef(std::string const& _varName, std::string const& _rhs);
+	void addCheckedVarDef(dataType _type, std::string const& _varName, std::string const& _rhs);
+	void appendTypedParams(
+			calleeType _calleeType,
+			bool _isValueType,
+			std::string const& _typeString,
+			std::string const& _varName,
+			delimiter _delimiter
+	);
+	void appendTypedParamsPublic(
+			bool _isValueType,
+			std::string const& _typeString,
+			std::string const& _varName,
+			delimiter _delimiter = delimiter::ADD
+	);
+	void appendTypedParamsExternal(
+			bool _isValueType,
+			std::string const& _typeString,
+			std::string const& _varName,
+			delimiter _delimiter = delimiter::ADD
+	);
+	void appendVarDeclToOutput(
+		std::string const& _type,
+		std::string const& _varName,
+		std::string const& _qualifier
+	);
+	void checkResizeOp(std::string const& _varName, unsigned _len);
+	void visitType(dataType _dataType, std::string const& _type, std::string const& _value);
+	void visitArrayType(std::string const&, ArrayType const&);
+	void createDeclAndParamList(
+		std::string const& _type,
+		dataType _dataType,
+		std::string& _varName
+	);
+	void appendVarDefToOutput(std::string const& _varName, std::string const& _rhs);
+	std::string equalityChecksAsString();
+	std::string typedParametersAsString(calleeType _calleeType);
+	void writeHelperFunctions();
+	void bufferVarDef(std::string const& _varName, std::string const& _rhs);
+
+	// Inline functions
+	inline unsigned getNextCounter()
+	{
+		return m_counter++;
+	}
+
+	inline std::string newVarName()
+	{
+		return  ("x_" + std::to_string(m_varCounter++));
+	}
+
+	inline std::string bytesArrayValueAsString()
+	{
+		return ("\"" + u256(h256(std::to_string(getNextCounter()))).str() + "\"");
```

Comment:
Can you explain the idea behind this construction?

Reply:
I updated this code to use keccak256 for hashing a counter value.